src/pool.rs

```diff
@@ -0,0 +1,35 @@
+use std::{sync, thread};
+
+pub(crate) type Task = dyn FnOnce() + Send;
+pub(crate) type BoxedTask = Box<Task>;
+
+pub(crate) struct ScopedPool<I>(I);
+
+impl<I> ScopedPool<I>
+where
+    I: IntoIterator<Item = BoxedTask>,
+    I::IntoIter: Send,
+{
+    pub(crate) fn new(tasks: I) -> Self {
+        Self(tasks)
+    }
+
+    pub(crate) fn run(self, test_threads: Option<usize>) {
+        let num_threads = test_threads
+            .or_else(|| thread::available_parallelism().ok().map(Into::into))
+            .unwrap_or(1);
+        let iter = sync::Mutex::new(self.0.into_iter());
+
+        thread::scope(|scope| {
+            let work = || {
+                let mut guard = iter.lock().unwrap();
```

Comment:
This is wrong: one worker thread will acquire the lock here and run all the tasks sequentially, while the other worker threads will be blocked until the end and then see an empty iterator. At least I'm 99% sure that's what will happen -- if there is a subtle reason why it actually works, I'd advocate for making the code less subtle!

The mechanical alternative is to only acquire the lock for the duration of each `next()` call. This requires desugaring the for-loop a bit. Perhaps extract the "get next task" bit into a dedicated closure and write the workers as `while let Some(task) = next_task() { task(); }`?

Reply:
I agree. Updated to what you propose here.